### PR TITLE
Allow easier modification of the OpenStreetMap-carto style

### PR DESCRIPTION
**Pull request by galewis2**
Now, the style isn't built into the image, but is now either hosted in a volume/local directory or worse case, fetched at runtime.

~It uses a wildcard to find critical style elements (like index.sql, lua file, etc) so there's some uncontrolled flexibility in different file names.~

Would replace #42 and #97 

~Still need to add some documentation, but here's the existing work.~

### CODE REVIEW

**Comment by galewis2** [comment]
This is based on the other PR I have, which improves the updating experience.

File: run.sh [hunk 1]
```diff
@@ -75,13 +91,14 @@ if [ "$1" = "import" ]; then
     fi
 
     # Import data
-    sudo -u renderer osm2pgsql -d gis --create --slim -G --hstore --tag-transform-script /home/renderer/src/openstreetmap-carto/openstreetmap-carto.lua --number-processes ${THREADS:-4} -S /home/renderer/src/openstreetmap-carto/openstreetmap-carto.style /data.osm.pbf ${OSM2PGSQL_EXTRA_ARGS:-}
+    sudo -u renderer osm2pgsql -d gis --create --slim -G --hstore --tag-transform-script /home/renderer/src/openstreetmap-carto/*.lua --number-processes ${THREADS:-4} -S /home/renderer/src/openstreetmap-carto/*.style /data.osm.pbf ${OSM2PGSQL_EXTRA_ARGS:-}
```
Comment (dbahrdt) [inline run.sh:94]: What happens if I have multiple styles/transform scripts?

**Comment by dbahrdt** [comment]
As general question:
How do we want to merge request? Do we squash the changes before merging? If not then I would strongly recommend to have useful commits. E.g. no "WIP" commits and the like.

**Comment by Overv** [comment]
@dbahrdt Yeah, I think all PRs should use squash merges as a general rule.

**Comment by galewis2** [comment]
@dbahrdt  Yea, I always do squash commits to keep main tidy. I except the dev branches to be polluted with non-useful commits.

File: run.sh (hunk 1 above)
Comment (galewis2) [reply run.sh:94]: I designed this to assume there's only one style/transform script, or just take the first one.

In hindsight, I will make a variable but give it defaults.

**Comment by galewis2** [comment]
> Hi, thank you for working on this very important modification. Is it possible to load a modified style created with [mapunik](https://maputnik.github.io) and therefore exported as a `json` style file?

@tommasodelorenzo  No, this Docker image is not vector-tile compatible. See https://github.com/openmaptiles/openmaptiles

File: run.sh (hunk 1 above)
Comment (galewis2) [reply run.sh]: Made improvements, now you can specify the style element names as necessary. See the readme

**Comment by tommasodelorenzo** [comment]
@galewis2 . Ok thanks.
I would bother you with an additional quick question: if I make a small change to my style, is there a simple way to apply those changes without re-running the `import`?
The fact is that when running the `import` I get errors such as 
```
+ sudo -u postgres createuser renderer
createuser: error: creation of new role failed: ERROR:  role "renderer" already exists
```
So I have to remove all data and re-run the `import`. This takes a lot of time.

File: run.sh [hunk 2]
```diff
@@ -75,13 +92,14 @@ if [ "$1" = "import" ]; then
     fi
 
     # Import data
-    sudo -u renderer osm2pgsql -d gis --create --slim -G --hstore --tag-transform-script /home/renderer/src/openstreetmap-carto/openstreetmap-carto.lua --number-processes ${THREADS:-4} -S /home/renderer/src/openstreetmap-carto/openstreetmap-carto.style /data.osm.pbf ${OSM2PGSQL_EXTRA_ARGS:-}
+    sudo -u renderer osm2pgsql -d gis --create --slim -G --hstore --tag-transform-script /home/renderer/src/openstreetmap-carto/${NAME_LUA:-openstreetmap-carto.lua} --number-processes ${THREADS:-4} -S /home/renderer/src/openstreetmap-carto/${NAME_STYLE:-openstreetmap-carto.style} /data.osm.pbf ${OSM2PGSQL_EXTRA_ARGS:-}
```
Comment (nielsole) [inline run.sh:95]: Not all styles use a lua script for transformation, do they?
Maybe we should omit using --tag-transform-script at all in that case?

**Comment by nielsole** [comment]
> So I have to remove all data and re-run the `import`. This takes a lot of time.

The lua script transforms the data during import. The indexes/views might be different which could lead to performance problems if the script is not rerun.
I guess if you just changed the mml you should be able to omit the import step. Either way: best to test your styles on a small dataset :)

File: run.sh [hunk 3]
```diff
@@ -20,11 +20,28 @@ if [ "$#" -ne 1 ]; then
...
+    echo "    NAME_LUA: name of .lua script to run as part of the style"
+    echo "    NAME_STYLE: name of the .style to use"
+    echo "    NAME_MML: name of the .mml file to render to mapnik.xml"
+    echo "    NAME_SQL: name of the .sql file to use"
     exit 1
 fi
 
 set -x
 
+if [ ! "$(ls -A /home/renderer/src/openstreetmap-carto)" ]; then
```
Comment (Istador) [inline run.sh:32]:
```sh
if [ ! -d /home/renderer/src/openstreetmap-carto ]; then
```

**Review by nielsole — COMMENTED** [review]
Great PR!
I haven't setup custom styling myself yet, so my assessments may be off.
I think this could be further improved:
* README: Where can I find other styles? (Just one other example would already be great)
* better support lack of lua script: https://github.com/cyclosm/cyclosm-cartocss-style
* Better support for multiple SQL scripts e.g.: https://github.com/OpenRailwayMap/OpenRailwayMap-CartoCSS/blob/master/SETUP.md

I am a big fan of incremental improvement, so maybe we can mark this as "beta"/experimental or something in the README and include 1-2 styles that are known to work and then merge it?

File: run.sh (hunk 3 above)
Comment (nielsole) [reply run.sh:32]:
```suggestion
if [ ! -d /home/renderer/src/openstreetmap-carto ]; then
```

File: run.sh (hunk 2 above)
Comment (galewis2) [reply run.sh:95]: This PR was created from using openstreetmap-carto changes, I don't have any experience with other styles so I've just assumed for now the only styles that will be supported are small incremental local changes on the existing style

File: Dockerfile [hunk 4]
```diff
@@ -115,9 +102,11 @@ RUN apt-get update \
  fonts-noto-hinted \
  fonts-noto-unhinted \
  gdal-bin \
+ git-core \
```
Comment (Istador) [inline Dockerfile:105]: Do we still need `git-core` in the final image?
Can we move `apt install git-core` and the `git clone` for `openstreetmap-carto` into its own stage (or does the `npm install -g carto@0.18.2` require it too)?

Comment (galewis2) [reply Dockerfile]: Resolved! I put the openstreetmap-carto fetch back into it's own stage, and just leveraged the git from a previous layer

File: run.sh (hunk 3 above)
Comment (galewis2) [reply run.sh:32]: I reverted this back as I need to confirm that openstreetmap-carto is empty, not just the directory exists (like you specify a bad path or something)

**Review by nielsole — APPROVED** [review]

**Review by nielsole — APPROVED** [review]
sorry for taking another 3 days for reapproval

**Pull request merged**